### PR TITLE
fix type errors in validator

### PR DESCRIPTION
**Pull request by benjstephenson**

### CODE REVIEW

File: lib/renderTo/index.test.ts [hunk 1]
```diff
@@ -211,7 +211,7 @@ test("[renderTo] allows deep nesting and increments headings", () => {
 																validation: {
 																	operand: 7,
 																	operation: "lessThan",
-																	returns: "number",
+																	returns: "boolean",
```
Comment (benjstephenson) [inline lib/renderTo/index.test.ts:214]: Ops defines this as returning a boolean.  I think I need to get back to writing the constructor helpers for the operation types so we don't need to specify the whole object when we create one.

File: lib/renderTo/index.ts [hunk 2]
```diff
@@ -46,7 +48,7 @@ const renderTo: RenderToF =
 		const elem = document.createElement(tagName === "HN" ? `H${lvl}` : tagName)
 
 		Object.entries(attributes).forEach(([attr, value]) =>
-			elem.setAttribute(attr, value),
+			elem.setAttribute(attr, String(value)),
```
Comment (benjstephenson) [inline lib/renderTo/index.ts:51]: value here is `string | boolean`, not sure if taking the string representation makes sense or not.

File: lib/renderTo/index.ts [hunk 3]
```diff
@@ -4,13 +4,15 @@ import type { TextNode } from "../types/shared"
...
+import { Option } from "@sitebender/fp/lib/option"
+import { CastableValue, Reify } from "@sitebender/operations/lib/types"
 
 type ValidatableElement = (
 	| HTMLInputElement
 	| HTMLSelectElement
 	| HTMLTextAreaElement
 ) & {
-	validate: (value: string) => void
+	validate: (value: Option<Reify<CastableValue>>) => void
```
Comment (benjstephenson) [inline lib/renderTo/index.ts:15]: I don't like this `Option<Reify<CastableValues>>` everywhere; I think operations could export a "validator" type that could hide this?

File: lib/renderTo/index.ts (hunk 2 above)
Comment (chasm) [reply lib/renderTo/index.ts:51]: It should (or will) be, I think, `string | number | boolean`, which is the reason for that. Some attributes are booleans and are typically input without a value:

```html
<input type="text" name="name" required>
```

If you look under the covers, this sets the value to "":

```html
<input type="text" name="name" required="">
```

There is no shortage of bad ideas in HTML. It grew in fits and starts with different people (or, gulp, _committees_) making decisions. So you have this sort of shit:

```ts
type Autocomplete = "true" | "false" | ""

type Required = boolean
```

We could leave out the `number` option, but I prefer this:

```json
{
  maxlength: 20
}
```

to this:

```json
{
  maxlength: "20"
}
```

Better for type checking, no?

File: lib/renderTo/index.ts (hunk 3 above)
Comment (chasm) [reply lib/renderTo/index.ts:15]: I am totally onboard with that. Definitely want a `ValidatorF` type.

**Comment by chasm** [comment]
Well, fun is over. Must start working in a branch now.

File: lib/renderTo/index.ts (hunk 2 above)
Comment (benjstephenson) [reply lib/renderTo/index.ts:51]: 100% prefer that too.

File: lib/renderTo/index.ts (hunk 3 above)
Comment (chasm) [reply lib/renderTo/index.ts:15]: Got type?

**Review by chasm — APPROVED** [review]

File: lib/renderTo/index.test.ts (hunk 1 above)
Comment (chasm) [reply lib/renderTo/index.test.ts:214]: Agreed. Yes. I was getting different errors here. No matter which one I chose, something told me it was the other.

But didn't we decide not to return booleans for these, but Some(value) where value is the value provided (the `operand` not the `test`)? That way the validator function passes in a value (in an Option?) and gets back either a Right(Some(value)) or a Left(Array(Error)).

Comment (benjstephenson) [reply lib/renderTo/index.test.ts:214]: I think if we do that we should change the naming of the operation, or add that behaviour as a separate operation? "Give me the smaller of these numbers" feels very different to "is this number smaller than that".  Plus naming it "lessThan" definitely feels like it overloads the meaning of "<" but changes the behaviour.

Comment (chasm) [reply lib/renderTo/index.test.ts:214]: Partly, this is because we're overloading the operations. For a calculation there is no difference between the operands -- all have equal weight. But for a validation, which does the same sorts of things, there is a big difference between the value being validated and all the rest.

Of course, some operations are clearly calculations. `add` must return the sum, not the initial operand.

But for those that are more validator-like, e.g., `lessThan`, it seems it should return the value tested.

Everything needs to return an Either -- we must be able to handle errors everywhere, validation or calculation. The question is, what goes in the either. Obviously, the left is a set of Errors (we might even consider Set for this to eliminate duplicates, although not sure how that works with objects).

But the Right? Internally we have Option so we can handle null/undefined. Is that always an error? If so, we don't need Option we just return a Left(Set(Error)). But if there are circumstances where a None is appropriate, then we should return Right(Some(value)).

No?

Maybe we need to inventory the various operations (list them on one page) and then think through them asking, _What should this return? Have we covered all edge cases?_ If none of them returns a None, then we can strip the Option out of our Either (OperationResult).

Comment (benjstephenson) [reply lib/renderTo/index.test.ts:214]: Listing the operations out will be helpful to decide if we do need the Option layer.
Do you think we'd want both the comparison and validator version of 'lessThan'?

Comment (benjstephenson) [reply lib/renderTo/index.test.ts:214]: what about 
```
lessThan: lessarOf
noLessThan: lessarOrEqualOf
noMoreThan: greaterOrEqualOf
moreThan: greaterOf 
```

Comment (chasm) [reply lib/renderTo/index.test.ts:214]: Um, _lesser_. This has the same problem as `min`, no? We don't want the lesser value, we want to know if the _operand_ is the lesser value and get it back if it is.

Comment (benjstephenson) [reply lib/renderTo/index.test.ts:214]: and if it's not then we get back a `none`?

Comment (chasm) [reply lib/renderTo/index.test.ts:214]: Am rethinking this (see long diatribe in Slack). These should, you are right, return a boolean.

**Comment by chasm** [comment]
Is this still valid or should we close without merging?

**Comment by benjstephenson** [comment]
I think we can close without merging
